Fails to build with GCC 12 (pre-release) on ppc64le

<!--
If you have a question rather than reporting a bug please go to https://forum.opencv.org where you get much faster responses.
If you need further assistance please read [How To Contribute](https://github.com/opencv/opencv/wiki/How_to_contribute).

This is a template helping you to create an issue which can be processed as quickly as possible. This is the bug reporting section for the OpenCV library.
-->

##### System information (version)
<!-- Example
- OpenCV => 4.2
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2017
-->

- OpenCV => 4.5.5
- Operating System / Platform => Fedora Linux Rawhide (development version), `ppc64le` (64-bit PowerPC, little-endian)
- Compiler => GCC 12.0.0 (pre-release)

##### Detailed description

GCC was just updated to a pre-release of 12.0.0 in Fedora Rawhide (the development version of Fedora Linux), and OpenCV and packages that use it are failing to compile on `ppc64le` due to a problem with AltiVec/VSX intrinsics.

The downstream bug report is [here](https://bugzilla.redhat.com/show_bug.cgi?id=2041573).

I don’t know if this is an OpenCV bug or a GCC 12 bug.

```
In file included from /builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/base.hpp:661,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core.hpp:53,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/utility.hpp:56,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/src/precomp.hpp:53,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/src/alloc.cpp:43:
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_double2 vec_cvf(const vec_float4&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:466:22: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  466 | { return fn2(vec_sldw(a, a, 1)); }
      |                      ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:468:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_4_2'
  468 | VSX_IMPL_CONV_EVEN_4_2(vec_double2, vec_float4, vec_cvf,  vec_cvfo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_dword2 vec_ctsl(const vec_float4&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:466:22: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  466 | { return fn2(vec_sldw(a, a, 1)); }
      |                      ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:472:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_4_2'
  472 | VSX_IMPL_CONV_EVEN_4_2(vec_dword2,  vec_float4, vec_ctsl, vec_ctslo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_udword2 vec_ctul(const vec_float4&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:466:22: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  466 | { return fn2(vec_sldw(a, a, 1)); }
      |                      ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:473:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_4_2'
  473 | VSX_IMPL_CONV_EVEN_4_2(vec_udword2, vec_float4, vec_ctul, vec_ctulo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_float4 vec_cvf(const vec_double2&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:479:20: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  479 |     return vec_sldw(v4, v4, 3);                  \
      |                    ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:482:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_2_4'
  482 | VSX_IMPL_CONV_EVEN_2_4(vec_float4, vec_double2, vec_cvf, vec_cvfo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_float4 vec_ctf(const vec_dword2&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:479:20: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  479 |     return vec_sldw(v4, v4, 3);                  \
      |                    ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:483:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_2_4'
  483 | VSX_IMPL_CONV_EVEN_2_4(vec_float4, vec_dword2,  vec_ctf, vec_ctfo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_float4 vec_ctf(const vec_udword2&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:479:20: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  479 |     return vec_sldw(v4, v4, 3);                  \
      |                    ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:484:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_2_4'
  484 | VSX_IMPL_CONV_EVEN_2_4(vec_float4, vec_udword2, vec_ctf, vec_ctfo)
      | ^~~~~~~~~~~~~~~~~~~~~~
In file included from /builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/base.hpp:661,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core.hpp:53,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/utility.hpp:56,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/src/precomp.hpp:53,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/src/algorithm.cpp:43:
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_double2 vec_cvf(const vec_float4&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:466:22: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  466 | { return fn2(vec_sldw(a, a, 1)); }
      |                      ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:468:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_4_2'
  468 | VSX_IMPL_CONV_EVEN_4_2(vec_double2, vec_float4, vec_cvf,  vec_cvfo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_dword2 vec_ctsl(const vec_float4&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:466:22: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  466 | { return fn2(vec_sldw(a, a, 1)); }
      |                      ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:472:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_4_2'
  472 | VSX_IMPL_CONV_EVEN_4_2(vec_dword2,  vec_float4, vec_ctsl, vec_ctslo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_udword2 vec_ctul(const vec_float4&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:466:22: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  466 | { return fn2(vec_sldw(a, a, 1)); }
      |                      ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:473:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_4_2'
  473 | VSX_IMPL_CONV_EVEN_4_2(vec_udword2, vec_float4, vec_ctul, vec_ctulo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_float4 vec_cvf(const vec_double2&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:479:20: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  479 |     return vec_sldw(v4, v4, 3);                  \
      |                    ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:482:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_2_4'
  482 | VSX_IMPL_CONV_EVEN_2_4(vec_float4, vec_double2, vec_cvf, vec_cvfo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_float4 vec_ctf(const vec_dword2&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:479:20: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  479 |     return vec_sldw(v4, v4, 3);                  \
      |                    ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:483:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_2_4'
  483 | VSX_IMPL_CONV_EVEN_2_4(vec_float4, vec_dword2,  vec_ctf, vec_ctfo)
      | ^~~~~~~~~~~~~~~~~~~~~~
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp: In function 'vec_float4 vec_ctf(const vec_udword2&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:479:20: error: invalid parameter combination for AltiVec intrinsic '__builtin_vec_sldw'
  479 |     return vec_sldw(v4, v4, 3);                  \
      |                    ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/vsx_utils.hpp:484:1: note: in expansion of macro 'VSX_IMPL_CONV_EVEN_2_4'
  484 | VSX_IMPL_CONV_EVEN_2_4(vec_float4, vec_udword2, vec_ctf, vec_ctfo)
      | ^~~~~~~~~~~~~~~~~~~~~~
```

##### Steps to reproduce

<!-- to add code example fence it with triple backticks and optional file extension
    ```.cpp
    // C++ code example
    ```
 or attach as .txt or .zip file
-->
Compile with GCC 12 on `ppc64le`.

##### Issue submission checklist

 - [x] I report the issue, it's not a question
   <!--
   OpenCV team works with forum.opencv.org, Stack Overflow and other communities
   to discuss problems. Tickets with question without real issue statement will be
   closed.
   -->
 - [x] I checked the problem with documentation, FAQ, open issues,
       forum.opencv.org, Stack Overflow, etc and have not found solution
   <!--
   Places to check:
   * OpenCV documentation: https://docs.opencv.org
   * FAQ page: https://github.com/opencv/opencv/wiki/FAQ
   * OpenCV forum: https://forum.opencv.org
   * OpenCV issue tracker: https://github.com/opencv/opencv/issues?q=is%3Aissue
   * Stack Overflow branch: https://stackoverflow.com/questions/tagged/opencv
   -->
 - [x] I updated to latest OpenCV version and the issue is still there
   <!--
   master branch for OpenCV 4.x and 3.4 branch for OpenCV 3.x releases.
   OpenCV team supports only latest release for each branch.
   The ticket is closed, if the problem is not reproduced with modern version.
   -->
 - [x] There is reproducer code and related data files: videos, images, onnx, etc *N/A*
   <!--
   The best reproducer -- test case for OpenCV that we can add to the library.
   Recommendations for media files and binary files:
   * Try to reproduce the issue with images and videos in opencv_extra repository
     to reduce attachment size
   * Use PNG for images, if you report some CV related bug, but not image reader
     issue
   * Attach the image as archive to the ticket, if you report some reader issue.
     Image hosting services compress images and it breaks the repro code.
   * Provide ONNX file for some public model or ONNX file with with random weights,
     if you report ONNX parsing or handling issue. Architecture details diagram
     from netron tool can be very useful too. See https://lutzroeder.github.io/netron/
   -->

The recent  updates of GCC12 (pre-release) bring this problem (again see https://github.com/opencv/opencv/issues/19020#issuecomment-1030940303 )

build are done with LTO disabled 

https://kojipkgs.fedoraproject.org/work/tasks/9006/82469006/build.log

```
[  1%] Building CXX object CMakeFiles/ade.dir/3rdparty/ade/ade-0.1.1f/sources/ade/source/passes/communications.cpp.o
/usr/bin/g++ -DVK_NO_PROTOTYPES -I/builddir/build/BUILD/opencv-4.5.5/redhat-linux-build -I/usr/include/vulkan -I/builddir/build/BUILD/opencv-4.5.5/redhat-linux-build/3rdparty/ade/ade-0.1.1f/sources/ade/include -isystem /usr/include/gdal -isystem /usr/include/coin -isystem /usr/include/openblas -O2  -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64 -mcpu=power8 -mtune=power8 -fasynchronous-unwind-tables -fstack-clash-protection   -fsigned-char -W -Wall -Werror=return-type -Werror=non-virtual-dtor -Werror=address -Werror=sequence-point -Wformat -Werror=format-security -Wmissing-declarations -Wundef -Winit-self -Wpointer-arith -Wshadow -Wsign-promo -Wuninitialized -Wsuggest-override -Wno-delete-non-virtual-dtor -Wno-comment -Wimplicit-fallthrough=3 -Wno-strict-overflow -fdiagnostics-show-option -pthread -fomit-frame-pointer -ffunction-sections -fdata-sections  -mcpu=power8 -fvisibility=hidden -fvisibility-inlines-hidden -fopenmp -fPIC -std=c++11 -MD -MT CMakeFiles/ade.dir/3rdparty/ade/ade-0.1.1f/sources/ade/source/passes/communications.cpp.o -MF CMakeFiles/ade.dir/3rdparty/ade/ade-0.1.1f/sources/ade/source/passes/communications.cpp.o.d -o CMakeFiles/ade.dir/3rdparty/ade/ade-0.1.1f/sources/ade/source/passes/communications.cpp.o -c /builddir/build/BUILD/opencv-4.5.5/redhat-linux-build/3rdparty/ade/ade-0.1.1f/sources/ade/source/passes/communications.cpp
In file included from /builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin.hpp:223,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/src/precomp.hpp:88,
                 from /builddir/build/BUILD/opencv-4.5.5/modules/core/src/algorithm.cpp:43:
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp: In function 'unsigned int cv::hal_baseline::v_reduce_sad(const v_uint8x16&, const v_uint8x16&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp:897:40: error: '__builtin_altivec_vadub' requires the '-mcpu=power9' and '-mvsx' options
  897 |     vec_uint4 sum4 = vec_sum4s(vec_absd(a.val, b.val), zero4);
      |                                        ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp:897:40: note: overloaded builtin '__builtin_vec_vadu' is implemented by builtin '__builtin_altivec_vadub'
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp: In function 'unsigned int cv::hal_baseline::v_reduce_sad(const v_uint16x8&, const v_uint16x8&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp:909:30: error: '__builtin_altivec_vaduh' requires the '-mcpu=power9' and '-mvsx' options
  909 |     vec_ushort8 ad = vec_absd(a.val, b.val);
      |                              ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp:909:30: note: overloaded builtin '__builtin_vec_vadu' is implemented by builtin '__builtin_altivec_vaduh'
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp: In function 'unsigned int cv::hal_baseline::v_reduce_sad(const v_uint32x4&, const v_uint32x4&)':
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp:922:34: error: '__builtin_altivec_vaduw' requires the '-mcpu=power9' and '-mvsx' options
  922 |     const vec_uint4 ad = vec_absd(a.val, b.val);
      |                                  ^
/builddir/build/BUILD/opencv-4.5.5/modules/core/include/opencv2/core/hal/intrin_vsx.hpp:922:34: note: overloaded builtin '__builtin_vec_vadu' is implemented by builtin '__builtin_altivec_vaduw'
```


 I also found this notes : 
On Wed, 2020-07-29 at 12:19 -0500, Steven Munroe wrote:
> Jeff Law wrote:
> 
> > For ppc64le is that the build was done with p8, but there is one
> > function (__builtin_altivec_vadub) that requires p9. This seems
> > like a package bug at first glance, not an LTO issue.
> 
> Yup __builtin_altivec_vadub is POWER9_VECTOR only.
> 
> For P8/6 use something like this:
> 
>   vmin = vec_min (a, b);
>   vmax = vec_max (a, b);
>   result = vec_sub (vmax, vmin);
> 
> executes in 4 cycles.

intrin_vsx.hpp:897:40: note: overloaded builtin '__builtin_vec_vadu' is implemented by builtin '__builtin_altivec_vadub' and
function (__builtin_altivec_vadub) requires p9 ...

--- a/modules/core/include/opencv2/core/vsx_utils.hpp
+++ b/modules/core/include/opencv2/core/vsx_utils.hpp
@@ -684,7 +684,8 @@ VSX_IMPL_LOAD_L8(vec_double2, double)
 #endif
 
 // absolute difference
-#ifndef vec_absd
+#ifndef _ARCH_PWR9
+#   undef vec_absd
 #   define vec_absd(a, b) vec_sub(vec_max(a, b), vec_min(a, b))
 #endif
 

